src/main/java/com/yahoo/sketches/theta/ConcurrentDirectThetaSketch.java

```diff
@@ -0,0 +1,317 @@
+  // affect the sketch at epoch j>i.
+  private volatile long epoch_;
+
+  /**
+   * Get a new sketch instance and initialize the given Memory as its backing store.
+   *
+   * @param lgNomLongs <a href="{@docRoot}/resources/dictionary.html#lgNomLongs">See lgNomLongs</a>.
+   * @param seed       <a href="{@docRoot}/resources/dictionary.html#seed">See Update Hash Seed</a>.
+   * @param dstMem     the given Memory object destination. It cannot be null.
+   *                   The hash table area will be cleared prior to use.
+   */
+  ConcurrentDirectThetaSketch(final int lgNomLongs, final long seed, final WritableMemory dstMem) {
+    super(lgNomLongs, seed, 1.0F, //p
+        ResizeFactor.X1, //rf,
+        null, dstMem, false);
+
+    volatileThetaLong_ = Long.MAX_VALUE;
+    volatileEstimate_ = 0;
+    sharedPropagationInProgress_ = new AtomicBoolean(false);
+    epoch_ = 0;
+  }
+
+  //Concurrent methods
+
+  /**
+   * Returns a (fresh) estimation of the number of unique entries
+   * @return a (fresh) estimation of the number of unique entries
+   */
+  @Override public double getEstimationSnapshot() {
+    return volatileEstimate_;
+  }
+
+  /**
+   * Updates the estimation of the number of unique entries by capturing a snapshot of the sketch
+   * data, namely, volatile theta and the num of valid entries in the sketch
+   */
+  @Override public void updateEstimationSnapshot() {
+    volatileEstimate_ = getEstimate();
+  }
+
+  /**
+   * Returns the value of the volatile theta manged by the shared sketch
+   * @return the value of the volatile theta manged by the shared sketch
+   */
+  @Override public long getVolatileTheta() {
+    return volatileThetaLong_;
+  }
+
+  /**
+   * Updates the value of the volatile theta by extracting it from the underlying sketch managed
+   * by the shared sketch
+   */
+  @Override public void updateVolatileTheta() {
+    volatileThetaLong_ = getThetaLong();
+  }
+
+  /**
+   * Returns true if a propagation is in progress, otherwise false
+   * @return an indication of whether there is a pending propagation in progress
+   */
+  @Override public boolean isPropagationInProgress() {
+    return sharedPropagationInProgress_.get();
+  }
+
+  /**
+   * Propagate the ConcurrentHeapThetaBuffer into this sketch
+   *
+   * @param localPropagationInProgress the flag to be updated when done
+   * @param sketchIn                   any Theta sketch with the data
+   * @param singleHash                 a single hash value
+   */
+  @Override public void propagate(final AtomicBoolean localPropagationInProgress,
+      final Sketch sketchIn, final long singleHash) {
+    final long epoch = epoch_;
+    final long k = 1 << getLgNomLongs();
+    if ((singleHash != NOT_SINGLE_HASH           // namely, is a single hash
+)
+        && (getRetainedEntries(false) < (2 * k)  // and a small sketch
+)
+    ) {                                         // then propagate myself (blocking)
+      startPropagation();
```

Comment:
This formatting looks weird.  The right parens need to be on the previous line. Please fix.